File: sdks/dart/lib/src/deserializer.dart
import 'dart:convert';
import 'dart:typed_data';
import 'constants.dart';
import 'errors.dart';
import 'uleb128.dart';

/// Cached constants for signed integer conversion
final BigInt _signBit64 = BigInt.one << 63;
final BigInt _signBit128 = BigInt.one << 127;
final BigInt _signBit256 = BigInt.one << 255;
final BigInt _mask64 = BigInt.one << 64;
final BigInt _mask128 = BigInt.one << 128;
final BigInt _mask256 = BigInt.one << 256;

/// BCS Deserializer - Manual deserialization API
class BcsDeserializer {
  /// Creates a deserializer from a Uint8List.
  BcsDeserializer(Uint8List data) : _data = data;

  /// Create from a list of integers
  factory BcsDeserializer.fromList(List<int> data) {
    return BcsDeserializer(Uint8List.fromList(data));
  }

  final Uint8List _data;
  int _offset = 0;
  int _depth = 0;

  // ==========================================================================
  // BOOLEAN
  // ==========================================================================

  /// Read a boolean value
  bool readBool() {
    if (_offset >= _data.length) {
      throw BcsError.unexpectedEof();
    }
    final byte = _data[_offset++];
    if (byte == 0) return false;
    if (byte == 1) return true;
    throw BcsError.invalidBoolean(byte);
  }

  // ==========================================================================
  // UNSIGNED INTEGERS
  // ==========================================================================

  /// Read an unsigned 8-bit integer
  int readU8() {
    if (_offset >= _data.length) {
      throw BcsError.unexpectedEof();
    }
    return _data[_offset++];
  }

  /// Read an unsigned 16-bit integer (little-endian)
  int readU16() {
    if (_offset + 2 > _data.length) {
      throw BcsError.unexpectedEof();
    }
    final value = _data[_offset] | (_data[_offset + 1] << 8);
    _offset += 2;
    return value;
  }

  /// Read an unsigned 32-bit integer (little-endian)
  int readU32() {
    if (_offset + 4 > _data.length) {
      throw BcsError.unexpectedEof();
    }
    // Unrolled for performance
    final value = _data[_offset] |
        (_data[_offset + 1] << 8) |
        (_data[_offset + 2] << 16) |
        (_data[_offset + 3] << 24);
    _offset += 4;
    return value;
  }

  /// Read an unsigned 64-bit integer (little-endian) as [BigInt].
  ///
  /// For values known to fit in a native [int], consider using [readU64Int]
  /// for better performance.
  BigInt readU64() {
    if (_offset + 8 > _data.length) {
      throw BcsError.unexpectedEof();
    }
    return _readBigIntLE(8);
  }

  /// Read an unsigned 64-bit integer (little-endian) as native [int].
  ///
  /// This is an optimized version of [readU64] that returns a native [int].
  /// On the Dart VM, this supports the full u64 range. On Dart Web (JavaScript),
  /// values larger than 2^53-1 may lose precision.
  ///
  /// For values that require full 64-bit precision on all platforms,
  /// use [readU64] which returns [BigInt].
  int readU64Int() {
    if (_offset + 8 > _data.length) {
      throw BcsError.unexpectedEof();
    }
    // Read directly as int for better performance
    final value = _data[_offset] |
        (_data[_offset + 1] << 8) |
        (_data[_offset + 2] << 16) |
        (_data[_offset + 3] << 24) |
        (_data[_offset + 4] << 32) |
        (_data[_offset + 5] << 40) |
        (_data[_offset + 6] << 48) |
        (_data[_offset + 7] << 56);
    _offset += 8;
    return value;
  }

  /// Read an unsigned 128-bit integer (little-endian)
  BigInt readU128() {
    if (_offset + 16 > _data.length) {
      throw BcsError.unexpectedEof();
    }
    return _readBigIntLE(16);
  }

  /// Read an unsigned 256-bit integer (little-endian)
  BigInt readU256() {
    if (_offset + 32 > _data.length) {
      throw BcsError.unexpectedEof();
    }
    return _readBigIntLE(32);
  }

  // ==========================================================================
  // SIGNED INTEGERS
  // ==========================================================================

  /// Read a signed 8-bit integer
  int readI8() {
    final value = readU8();
    return value >= 0x80 ? value - 0x100 : value;
  }

  /// Read a signed 16-bit integer (little-endian)
  int readI16() {
    final value = readU16();
    return value >= 0x8000 ? value - 0x10000 : value;
  }

  /// Read a signed 32-bit integer (little-endian)
  int readI32() {
    final value = readU32();
    return value >= 0x80000000 ? value - 0x100000000 : value;
  }

  /// Read a signed 64-bit integer (little-endian)
  BigInt readI64() {
    final value = readU64();
    return value >= _signBit64 ? value - _mask64 : value;
  }

  /// Read a signed 128-bit integer (little-endian)
  BigInt readI128() {
    final value = readU128();
    return value >= _signBit128 ? value - _mask128 : value;
  }

  /// Read a signed 256-bit integer (little-endian)
  BigInt readI256() {
    final value = readU256();
    return value >= _signBit256 ? value - _mask256 : value;
  }

  // ==========================================================================
  // ULEB128
  // ==========================================================================

  /// Read a ULEB128-encoded value
  int readUleb128() {
    final result = Uleb128.decode(_data, _offset);
    _offset += result.bytesRead;
    return result.value;
  }

  // ==========================================================================
  // BYTES AND STRINGS
  // ==========================================================================

  /// Read fixed-length bytes (without length prefix) as a copy.
  ///
  /// Returns a new [Uint8List] containing a copy of the bytes.
  /// This is safe to modify without affecting the deserializer.
  ///
  /// For performance-critical code where you don't need to modify the bytes,
  /// consider using [readFixedBytesView] instead.
  Uint8List readFixedBytes(int length) {
    if (_offset + length > _data.length) {
      throw BcsError.unexpectedEof();
    }
    final result = Uint8List.fromList(_data.sublist(_offset, _offset + length));
    _offset += length;
    return result;
  }

  /// Read fixed-length bytes (without length prefix) as a view (zero-copy).
  ///
  /// **WARNING:** The returned [Uint8List] is a view into the input buffer.
  /// Modifying the returned bytes will modify the original input data.
  /// Only use this when you need maximum performance and won't modify the result.
  Uint8List readFixedBytesView(int length) {
    if (_offset + length > _data.length) {
      throw BcsError.unexpectedEof();
    }
    final result = Uint8List.sublistView(_data, _offset, _offset + length);
    _offset += length;
    return result;
  }

  /// Read bytes with ULEB128 length prefix as a copy.
  ///
  /// Returns a new [Uint8List] containing a copy of the bytes.
  Uint8List readBytes() {
    final length = readUleb128();
    _checkSequenceLength(length);
    return readFixedBytes(length);
  }

  /// Read bytes with ULEB128 length prefix as a view (zero-copy).
  ///
  /// **WARNING:** The returned [Uint8List] is a view into the input buffer.
  Uint8List readBytesView() {
    final length = readUleb128();
    _checkSequenceLength(length);
    return readFixedBytesView(length);
  }

  /// Read a UTF-8 string with ULEB128 length prefix
  String readString() {
    final length = readUleb128();
    _checkSequenceLength(length);
    if (_offset + length > _data.length) {
      throw BcsError.unexpectedEof();
    }
    // Decode directly from view to avoid copy (utf8.decode doesn't modify input)
    try {
      final str = utf8.decode(
        Uint8List.sublistView(_data, _offset, _offset + length),
        allowMalformed: false,
      );
      _offset += length;
      return str;
    } on FormatException {
      throw BcsError.invalidUtf8();
    }
  }

  // ==========================================================================
  // COMPOSITE TYPES
  // ==========================================================================

  /// Read an optional value
  T? readOption<T>(T Function(BcsDeserializer) deserializer) {
    final tag = readU8();
    switch (tag) {
      case 0:
        return null;
      case 1:
        return deserializer(this);
      default:
        throw BcsError.invalidOption(tag);
    }
  }

  /// Read a vector with element deserializer
  List<T> readVector<T>(T Function(BcsDeserializer) deserializer) {
    final length = readUleb128();
    _checkSequenceLength(length);
    // Pre-allocate list with known size for better performance
    final result = List<T>.filled(length, null as T);
    for (var i = 0; i < length; i++) {
      result[i] = deserializer(this);
    }
    return result;
  }

  /// Read a map with key/value deserializers
  Map<K, V> readMap<K, V>(
    K Function(BcsDeserializer) keyDeserializer,
    V Function(BcsDeserializer) valueDeserializer,
  ) {
    final length = readUleb128();
    _checkSequenceLength(length);

    final result = <K, V>{};
    Uint8List? prevKeyBytes;

    for (var i = 0; i < length; i++) {
      // Record start position to get key bytes
      final keyStart = _offset;

      final key = keyDeserializer(this);

      // Get key bytes for ordering check (use view - we only compare, don't modify)
      final keyBytes = Uint8List.sublistView(_data, keyStart, _offset);

      // Check ordering
      if (prevKeyBytes != null) {
        final cmp = _compareBytes(keyBytes, prevKeyBytes);
        if (cmp <= 0) {
          if (cmp == 0) {
            throw BcsError.duplicateMapKey();
          }
          throw BcsError.nonCanonicalMap();
        }
      }
      // Store a copy for the next comparison since we'll reuse the view
      prevKeyBytes = Uint8List.fromList(keyBytes);

      final value = valueDeserializer(this);
      result[key] = value;
    }

    return result;
  }

  /// Read an enum variant index (ULEB128)
  int readVariantIndex() {
    _enterContainer();
    return readUleb128();
  }

  // ==========================================================================
  // CONTAINER DEPTH
  // ==========================================================================

  /// Enter a struct container for depth tracking
  BcsDeserializer enterStruct([String name = '']) {
    _enterContainer();
    return this;
  }

  /// Leave the current struct container
  BcsDeserializer leaveStruct() {
    _leaveContainer();
    return this;
  }

  /// Enter an enum container and read variant index
  int enterEnum() {
    return readVariantIndex();
  }

  /// Leave the current enum container
  BcsDeserializer leaveEnum() {
    _leaveContainer();
    return this;
  }

  // ==========================================================================
  // STATE
  // ==========================================================================

  /// Check that all input has been consumed
  void checkEnd() {
    if (_offset < _data.length) {
      throw BcsError.remainingInput(_data.length - _offset);
    }
  }

  /// Get the current offset
  int get offset => _offset;

  /// Get the remaining bytes count
  int get remaining => _data.length - _offset;

  /// Check if there's more data to read
  bool get hasRemaining => _offset < _data.length;

  // ==========================================================================
  // PRIVATE HELPERS
  // ==========================================================================

  void _checkSequenceLength(int length) {
    if (length > maxSequenceLength) {
      throw BcsError.exceededMaxLength(length, maxSequenceLength);
    }
  }

  void _enterContainer() {
    _depth++;
    if (_depth > maxContainerDepth) {
      throw BcsError.exceededContainerDepth(_depth, maxContainerDepth);
    }
  }

  void _leaveContainer() {
    if (_depth > 0) {
      _depth--;
    }
  }

  /// Optimized BigInt read - builds value from bytes.
  ///
  /// Uses 64-bit chunks for better performance with large integers (u128, u256).
  /// This reduces the number of BigInt operations from O(n) to O(n/8).
  BigInt _readBigIntLE(int byteLength) {
    var value = BigInt.zero;

    // Process in 64-bit chunks for better performance
    final fullChunks = byteLength ~/ 8;
    final remainingBytes = byteLength % 8;

    // Process full 64-bit chunks from most significant to least significant
    for (var chunk = fullChunks - 1; chunk >= 0; chunk--) {
      final chunkOffset = _offset + chunk * 8;
      // Build 64-bit word from little-endian bytes
      final word = _data[chunkOffset] |
          (_data[chunkOffset + 1] << 8) |
          (_data[chunkOffset + 2] << 16) |
          (_data[chunkOffset + 3] << 24) |
          (_data[chunkOffset + 4] << 32) |
          (_data[chunkOffset + 5] << 40) |
          (_data[chunkOffset + 6] << 48) |
          (_data[chunkOffset + 7] << 56);

      if (value == BigInt.zero) {
        value = BigInt.from(word).toUnsigned(64);
      } else {
        value = (value << 64) | BigInt.from(word).toUnsigned(64);
      }
    }

    // Handle any remaining bytes (for non-8-byte-aligned lengths)
    if (remainingBytes > 0) {
      final remainingOffset = _offset + fullChunks * 8;
      var remaining = 0;
      for (var i = remainingBytes - 1; i >= 0; i--) {
        remaining = (remaining << 8) | _data[remainingOffset + i];
      }
      if (value == BigInt.zero) {
        value = BigInt.from(remaining);
      } else {
        value = (value << (remainingBytes * 8)) | BigInt.from(remaining);
      }
    }

    _offset += byteLength;
    return value;
  }

  static int _compareBytes(Uint8List a, Uint8List b) {
    final minLen = a.length < b.length ? a.length : b.length;
    for (var i = 0; i < minLen; i++) {
      if (a[i] < b[i]) return -1;
      if (a[i] > b[i]) return 1;
    }
    return a.length.compareTo(b.length);
  }
}

File: sdks/typescript/src/deserializer.ts
/**
 * BCS Deserializer - Manual deserialization API.
 *
 * Provides explicit methods for deserializing each BCS type.
 * Use this for full control over deserialization.
 *
 * @example
 * ```ts
 * const data = new Uint8Array([1, 100, 0, 0, 0, 0, 0, 0, 0, 5, 104, 101, 108, 108, 111]);
 * const des = new BcsDeserializer(data);
 * const value1 = des.readU8();    // 1
 * const value2 = des.readU64();   // 100n
 * const value3 = des.readString(); // "hello"
 * des.checkEnd(); // Verify no remaining bytes
 * ```
 */

import { BcsError } from "./errors";
import * as uleb128 from "./uleb128";
import { MAX_SEQUENCE_LENGTH } from "./serializer";

// Cached TextDecoder for string deserialization
const textDecoder = new TextDecoder("utf-8", { fatal: true });

// Cached sign bit constants
const SIGN_BIT_64 = 1n << 63n;
const SIGN_BIT_128 = 1n << 127n;
const SIGN_BIT_256 = 1n << 255n;
const MASK_64 = 1n << 64n;
const MASK_128 = 1n << 128n;
const MASK_256 = 1n << 256n;

// Container depth limit
const MAX_CONTAINER_DEPTH = 500;

/**
 * BCS Deserializer class.
 */
export class BcsDeserializer {
  private data: Uint8Array;
  private offset: number = 0;
  private depth: number = 0;

  /**
   * Create a new deserializer.
   * @param data - The data to deserialize from
   */
  constructor(data: Uint8Array) {
    this.data = data;
  }

  // ==========================================================================
  // BOOLEAN
  // ==========================================================================

  /**
   * Deserialize a boolean value.
   */
  readBool(): boolean {
    if (this.offset >= this.data.length) {
      throw BcsError.unexpectedEof(1, 0);
    }
    const byte = this.data[this.offset++];
    if (byte === 0) return false;
    if (byte === 1) return true;
    throw BcsError.invalidBoolean(byte);
  }

  // ==========================================================================
  // UNSIGNED INTEGERS
  // ==========================================================================

  /**
   * Deserialize an unsigned 8-bit integer.
   */
  readU8(): number {
    if (this.offset >= this.data.length) {
      throw BcsError.unexpectedEof(1, 0);
    }
    return this.data[this.offset++];
  }

  /**
   * Deserialize an unsigned 16-bit integer (little-endian).
   */
  readU16(): number {
    if (this.offset + 2 > this.data.length) {
      throw BcsError.unexpectedEof(2, this.data.length - this.offset);
    }
    const value = this.data[this.offset] | (this.data[this.offset + 1] << 8);
    this.offset += 2;
    return value;
  }

  /**
   * Deserialize an unsigned 32-bit integer (little-endian).
   */
  readU32(): number {
    if (this.offset + 4 > this.data.length) {
      throw BcsError.unexpectedEof(4, this.data.length - this.offset);
    }
    // Unrolled for performance
    const value =
      this.data[this.offset] |
      (this.data[this.offset + 1] << 8) |
      (this.data[this.offset + 2] << 16) |
      ((this.data[this.offset + 3] << 24) >>> 0);
    this.offset += 4;
    return value >>> 0; // Ensure unsigned
  }

  /**
   * Deserialize an unsigned 64-bit integer (little-endian).
   */
  readU64(): bigint {
    if (this.offset + 8 > this.data.length) {
      throw BcsError.unexpectedEof(8, this.data.length - this.offset);
    }
    return this.readBigIntLE(8);
  }

  /**
   * Deserialize an unsigned 128-bit integer (little-endian).
   */
  readU128(): bigint {
    if (this.offset + 16 > this.data.length) {
      throw BcsError.unexpectedEof(16, this.data.length - this.offset);
    }
    return this.readBigIntLE(16);
  }

  /**
   * Deserialize an unsigned 256-bit integer (little-endian).
   */
  readU256(): bigint {
    if (this.offset + 32 > this.data.length) {
      throw BcsError.unexpectedEof(32, this.data.length - this.offset);
    }
    return this.readBigIntLE(32);
  }

  // ==========================================================================
  // SIGNED INTEGERS
  // ==========================================================================

  /**
   * Deserialize a signed 8-bit integer (two's complement).
   */
  readI8(): number {
    const unsigned = this.readU8();
    return unsigned >= 128 ? unsigned - 256 : unsigned;
  }

  /**
   * Deserialize a signed 16-bit integer (two's complement, little-endian).
   */
  readI16(): number {
    const unsigned = this.readU16();
    return unsigned >= 0x8000 ? unsigned - 0x10000 : unsigned;
  }

  /**
   * Deserialize a signed 32-bit integer (two's complement, little-endian).
   */
  readI32(): number {
    const unsigned = this.readU32();
    return unsigned >= 0x80000000 ? unsigned - 0x100000000 : unsigned;
  }

  /**
   * Deserialize a signed 64-bit integer (two's complement, little-endian).
   */
  readI64(): bigint {
    const unsigned = this.readU64();
    return unsigned >= SIGN_BIT_64 ? unsigned - MASK_64 : unsigned;
  }

  /**
   * Deserialize a signed 128-bit integer (two's complement, little-endian).
   */
  readI128(): bigint {
    const unsigned = this.readU128();
    return unsigned >= SIGN_BIT_128 ? unsigned - MASK_128 : unsigned;
  }

  /**
   * Deserialize a signed 256-bit integer (two's complement, little-endian).
   */
  readI256(): bigint {
    const unsigned = this.readU256();
    return unsigned >= SIGN_BIT_256 ? unsigned - MASK_256 : unsigned;
  }

  // ==========================================================================
  // ULEB128
  // ==========================================================================

  /**
   * Deserialize a ULEB128-encoded unsigned integer.
   */
  readUleb128(): number {
    const [value, bytesRead] = uleb128.decode(this.data, this.offset);
    this.offset += bytesRead;
    return value;
  }

  // ==========================================================================
  // BYTES AND STRINGS
  // ==========================================================================

  /**
   * Deserialize a byte array (length-prefixed with ULEB128).
   */
  readBytes(): Uint8Array {
    const length = this.readUleb128();
    if (length > MAX_SEQUENCE_LENGTH) {
      throw BcsError.exceededMaxLength(length);
    }
    return this.readFixedBytes(length);
  }

  /**
   * Deserialize a UTF-8 string (length-prefixed with ULEB128).
   */
  readString(): string {
    const length = this.readUleb128();
    if (length > MAX_SEQUENCE_LENGTH) {
      throw BcsError.exceededMaxLength(length);
    }
    if (this.offset + length > this.data.length) {
      throw BcsError.unexpectedEof(length, this.data.length - this.offset);
    }
    // Use subarray for zero-copy view
    const bytes = this.data.subarray(this.offset, this.offset + length);
    try {
      const str = textDecoder.decode(bytes);
      this.offset += length;
      return str;
    } catch {
      throw BcsError.invalidUtf8();
    }
  }

  /**
   * Deserialize fixed-length bytes (no length prefix).
   * Returns a copy of the bytes.
   */
  readFixedBytes(length: number): Uint8Array {
    if (this.offset + length > this.data.length) {
      throw BcsError.unexpectedEof(length, this.data.length - this.offset);
    }
    // Use slice to create a copy (maintains existing behavior)
    const bytes = this.data.slice(this.offset, this.offset + length);
    this.offset += length;
    return bytes;
  }

  /**
   * Deserialize fixed-length bytes as a view (no copy, no length prefix).
   * WARNING: The returned view is only valid while the deserializer's data is not modified.
   */
  readFixedBytesView(length: number): Uint8Array {
    if (this.offset + length > this.data.length) {
      throw BcsError.unexpectedEof(length, this.data.length - this.offset);
    }
    // Use subarray for zero-copy view
    const bytes = this.data.subarray(this.offset, this.offset + length);
    this.offset += length;
    return bytes;
  }

  // ==========================================================================
  // OPTION
  // ==========================================================================

  /**
   * Deserialize an optional value.
   *
   * @param deserializer - Function to deserialize the inner value
   * @returns The value or null if None
   */
  readOption<T>(deserializer: (des: BcsDeserializer) => T): T | null {
    if (this.offset >= this.data.length) {
      throw BcsError.unexpectedEof(1, 0);
    }
    const tag = this.data[this.offset++];
    if (tag === 0) return null;
    if (tag === 1) return deserializer(this);
    throw BcsError.invalidOption(tag);
  }

  // ==========================================================================
  // VECTOR
  // ==========================================================================

  /**
   * Deserialize a vector of values.
   *
   * @param deserializer - Function to deserialize each element
   */
  readVector<T>(deserializer: (des: BcsDeserializer) => T): T[] {
    const length = this.readUleb128();
    // Validate length is within safe bounds
    if (length > MAX_SEQUENCE_LENGTH || !Number.isSafeInteger(length)) {
      throw BcsError.exceededMaxLength(length);
    }
    // Pre-allocate array for better performance
    const result: T[] = new Array(length);
    for (let i = 0; i < length; i++) {
      result[i] = deserializer(this);
    }
    return result;
  }

  // ==========================================================================
  // ENUM
  // ==========================================================================

  /**
   * Read an enum variant index (ULEB128).
   */
  readVariantIndex(): number {
    return this.readUleb128();
  }

  // ==========================================================================
  // CONTAINER DEPTH
  // ==========================================================================

  /**
   * Enter a struct container for depth tracking.
   * @throws BcsError if container depth exceeds MAX_CONTAINER_DEPTH (500)
   */
  enterStruct(): this {
    this.enterContainer("struct");
    return this;
  }

  /**
   * Leave a struct container.
   */
  leaveStruct(): this {
    this.leaveContainer();
    return this;
  }

  /**
   * Enter an enum container for depth tracking and read variant index.
   * @returns The enum variant index
   * @throws BcsError if container depth exceeds MAX_CONTAINER_DEPTH (500)
   */
  enterEnum(): number {
    this.enterContainer("enum");
    return this.readVariantIndex();
  }

  /**
   * Leave an enum container.
   */
  leaveEnum(): this {
    this.leaveContainer();
    return this;
  }

  /**
   * Get the current container depth.
   */
  get containerDepth(): number {
    return this.depth;
  }

  private enterContainer(name: string): void {
    if (this.depth >= MAX_CONTAINER_DEPTH) {
      throw BcsError.exceededContainerDepth(name);
    }
    this.depth++;
  }

  private leaveContainer(): void {
    if (this.depth > 0) {
      this.depth--;
    }
  }

  // ==========================================================================
  // MAP
  // ==========================================================================

  /**
   * Deserialize a map (verifying sorted keys).
   *
   * @param keyDeserializer - Function to deserialize keys
   * @param valueDeserializer - Function to deserialize values
   */
  readMap<K, V>(
    keyDeserializer: (des: BcsDeserializer) => K,
    valueDeserializer: (des: BcsDeserializer) => V
  ): Map<K, V> {
    const length = this.readUleb128();
    if (length > MAX_SEQUENCE_LENGTH) {
      throw BcsError.exceededMaxLength(length);
    }

    const result = new Map<K, V>();
    let prevKeyBytes: Uint8Array | null = null;

    for (let i = 0; i < length; i++) {
      // Record position before reading key
      const keyStart = this.offset;
      const key = keyDeserializer(this);
      const keyEnd = this.offset;
      // Use subarray for zero-copy view (only used for comparison)
      const keyBytes = this.data.subarray(keyStart, keyEnd);

      // Verify key order
      if (prevKeyBytes !== null) {
        const cmp = compareBytes(prevKeyBytes, keyBytes);
        if (cmp === 0) {
          throw BcsError.nonCanonicalMap("duplicate key");
        }
        if (cmp > 0) {
          throw BcsError.nonCanonicalMap("keys not sorted");
        }
      }
      // Copy key bytes only when storing for next comparison
      prevKeyBytes = this.data.slice(keyStart, keyEnd);

      const value = valueDeserializer(this);
      result.set(key, value);
    }

    return result;
  }

  // ==========================================================================
  // UTILITY
  // ==========================================================================

  /**
   * Check that all input has been consumed.
   * @throws BcsError if there is remaining input
   */
  checkEnd(): void {
    if (this.offset < this.data.length) {
      throw BcsError.remainingInput(this.data.length - this.offset);
    }
  }

  /**
   * Get the number of remaining bytes.
   */
  get remaining(): number {
    return this.data.length - this.offset;
  }

  /**
   * Get the current offset.
   */
  get position(): number {
    return this.offset;
  }

  // ==========================================================================
  // PRIVATE HELPERS
  // ==========================================================================

  /** Optimized BigInt read - processes from high to low byte for better performance */
  private readBigIntLE(byteLength: number): bigint {
    let value = 0n;
    for (let i = byteLength - 1; i >= 0; i--) {
      value = (value << 8n) | BigInt(this.data[this.offset + i]);
    }
    this.offset += byteLength;
    return value;
  }
}

/**
 * Compare two byte arrays lexicographically.
 */
function compareBytes(a: Uint8Array, b: Uint8Array): number {
  const minLen = Math.min(a.length, b.length);
  for (let i = 0; i < minLen; i++) {
    if (a[i] !== b[i]) {
      return a[i] - b[i];
    }
  }
  return a.length - b.length;
}
